### ISSUE TITLE
Send selection to upper/lower layer

### ISSUE THREAD

**Issue by Gnumaru** [bjorn/tiled#1415]
It would be useful to be capable of selecting tiles and sending them to upper/lower layers. I searched the opened feature requests but it seems none of them are about this subject.

Use Case Description:

1. The user "Foo" is making a map in Tiled.
2. Foo has drawn, in just one layer, a map consisting of both terrain and enemy mockups.
3. Foo has finished drawing the map but has now decided he wants to put the enemy mockups into a separate layer
4. Foo adds a new tile layer to the map
5. Foo selects all the enemies on the first (and now lower) layer using the "Select Same Tile" tool
6. Under the Edit menu, Foo selects the "Send selection to upper layer"
7. The application then moves the selection to the upper layer

Aditional Rules:

- If there is no upper or lower layer to send the selection to, the application fails silently and does nothing
- If the destination layer is not of the same type of the source layer, the application continues to search the next layers in order for one of the same type of the source layer. If none is found, de application fails silently and does nothing
- if there is nothing selected in the current layer, the "Send selection to upper layer" and "Send selection to lower layer" menu itens are disabled

**Comment by bjorn** [comment]
If the point is to make a new layer from the selection, there will already be an action for this in Tiled 0.18 (and is already available in the snapshot builds):

1. Foo selects all the enemies on the first (and now lower) layer using the "Select Same Tile" tool
2. Foo triggers _Layer -> New -> Layer via Cut_

Sending the selection to an arbitrary target layer, or moving it a layer up/down, could be useful as well but it would also be a bit dangerous because it will overwrite whatever content is already there at the target layer (so it could be quite destructive when used repeatedly to move a selection several layers up).

**Comment by Gnumaru** [comment]
I did not know about the "New Layer via Cut" feature in 0.18. But moving a selection between existing layers could still be useful in cases where the user accidentally works on a layer when in fact he was intending to work on another.

For example:

1. Foo is drawing a map with two layers of the "tile" type, one for terrain graphics and another for enemy mockups
2. The application has the enemies layer selected
3. Foo has accidentally draw a section of enemies and plaforms on the current layer (the enemies one).
4. Foo makes a large selection of the mixed-content map section with the "Rectangular Select" tool
5. Foo sends the selection to the terrain layer
6. Foo swiches to the terrain layer
7. Foo selects the enemies using the "Select Same Tile" tool
8. Foo sends the selection to the enemies layer

When moving tiles between layers I usually cut and paste the tiles. But if one have already drawn a complex and large map, and realized too late he was working on the wrong layer, it can be difficult/time consuming to move the tiles to another layer. When you select+cut+paste tiles, the mouse cursor is centered on the bounding rectagle of the original selection area, and it is likely that the mouse cursor could be far away from any of the selected tiles, so one would have to zoom out enough to see one of the selected tiles and position it on the right place. But since the view is zoomed out, it become difficult to place the tiles on the right position.


Maybe a simpler solution to this problem of moving tiles between existing layers could be a "Paste directly" feature _[ps.: please read the ps. at the end]_. Different from the default paste behaviour that effectively creates a temporary brush, the "Paste directly" could paste the contents of the clipboard directly in the current layer without creating a brush, in the same position they where when cutted/copied. This would override any content already existing in the pasted position, but this is what would happen with any paste action or brush use anyway.

The use case would be:

1. Foo wants to move a complex selection to another existing layer
2. Foo creates the complex selection using all the application features for making selections
3. Foo cuts the selection
4. Foo changes the current layer to the one he wishes to paste on
5. Under the Edit menu, Foo selects the "Paste directly" entry
6. The clipboard contents get pasted on the same position the where on the layer where they cutted/copied from.


Ps.:
While writing this, I have just realized that Tiled alterady has a "Paste in Place" action that apparently should do exactly what I described above. Unfortunately, I tried using it and it didn't worked =(. when cutting/copying and then using "paste in place" (ctrl+shift+v), the behaviour is the same of the normal paste, the clipboard contents became a temporary brush pattern that can be used as a normal brush. Imediatly after pressing ctrl+shift+v and keeping both ctrl and shift pressed, nothing happens, but when I release either ctrl or shift, then the default paste behaviour happens. I realized also that there is already one closed issue, "Paste in place #1257", about this bug. Maybe the defective behaviour is a regression bug?

I am using Tiled version 0.17.2 x64 on windows 10 x64 version 1511 compilation 10586.589.

**Comment by bjorn** [comment]
The "Paste in Place" action was requested for object selection rather than tile selection, and since the two work very differently it was only implemented for object selection. But you're right that the action should behave like you expected for tile selection.

**Comment by bjorn** [comment]
I've added this to the Tiled 0.18 milestone, as a reminder to look into implementing "Paste in Place" for tile selection.

**Comment by Gnumaru** [comment]
I see... so "Paste in Place" was intended for object layers only.

I'm glad to hear you consider implementing it for tile layers in a future version. I think it would be really useful.

**Comment by bjorn** [comment]
Please try whether the improved "Paste in Place" action provides a good solution to the use-cases presented. It will be available in Tiled 0.18 soon, as well as in the snapshot builds that are currently updating.

**Comment by Gnumaru** [comment]
Yes, it worked as expected.

I cloned the repository, built it with qtcreator and made a test as described in the previous comments.

1. I made a copy of one of my maps (just to be safe)
2. I opened the map on Tiled
3. I selected a layer wich had both terrain tiles and enemy tiles on the same layer
4. I selected the enemies using the "Select Same Tile" tool
5. I cutted the enemies with ctrl+x
6. I switched to the enemies layer
7. I pasted the content with ctrl+shift+v

I made the test again triggering the action via menu entry instead of the keyboard shortcut, and it worked as well.

Thanks for the update =)
